python/ray/tests/test_autoscaling_policy.py

```diff
@@ -0,0 +1,354 @@
+            event = self.event_queue.get()
+            self.process_event(event)
+        self.schedule()
+        return self.virtual_time
+
+
+SAMPLE_CLUSTER_CONFIG = copy.deepcopy(MULTI_WORKER_CLUSTER)
+SAMPLE_CLUSTER_CONFIG["min_workers"] = 0
+SAMPLE_CLUSTER_CONFIG["max_workers"] = 9999
+SAMPLE_CLUSTER_CONFIG["target_utilization_fraction"] = 1.0
+SAMPLE_CLUSTER_CONFIG["available_node_types"]["m4.16xlarge"]["max_workers"] = 100
+SAMPLE_CLUSTER_CONFIG["available_node_types"]["m4.4xlarge"]["max_workers"] = 10000
+
+
+class AutoscalingPolicyTest(unittest.TestCase):
+    def setUp(self):
+        _NODE_PROVIDERS["mock"] = lambda config: self.create_provider
+        self.provider = None
+        self.tmpdir = tempfile.mkdtemp()
+
+    def tearDown(self):
+        self.provider = None
+        del _NODE_PROVIDERS["mock"]
+        _clear_provider_cache()
+        shutil.rmtree(self.tmpdir)
+        ray.shutdown()
+
+    def waitForNodes(self, expected, comparison=None, tag_filters={}):
+        MAX_ITER = 50
+        for i in range(MAX_ITER):
+            n = len(self.provider.non_terminated_nodes(tag_filters))
+            if comparison is None:
+                comparison = self.assertEqual
+            try:
+                comparison(n, expected)
+                return
+            except Exception:
+                if i == MAX_ITER - 1:
+                    raise
+            time.sleep(0.1)
+
+    def create_provider(self, config, cluster_name):
+        assert self.provider
+        return self.provider
+
+    def write_config(self, config):
+        path = self.tmpdir + "/simple.yaml"
+        with open(path, "w") as f:
+            f.write(yaml.dump(config))
+        return path
+
+    def testManyTasks(self):
+        config = copy.deepcopy(SAMPLE_CLUSTER_CONFIG)
+        config_path = self.write_config(config)
+        self.provider = MockProvider()
+        simulator = Simulator(config_path, self.provider)
+
+        done_count = 0
+        def done_callback():
+            nonlocal done_count
+            done_count += 1
+
+        tasks = [Task(duration=200, resources={"CPU": 1}, done_callback=done_callback) for _ in range(10000)]
+        simulator.submit(tasks)
+
+        time = 0
+        while done_count < len(tasks):
+            time = simulator.step()
+
+        assert time < 200
+
+    def testManyActors(self):
+        config = copy.deepcopy(SAMPLE_CLUSTER_CONFIG)
+        config_path = self.write_config(config)
+        self.provider = MockProvider()
+        simulator = Simulator(config_path, self.provider)
+
+        start_count = 0
+        def start_callback():
+            nonlocal start_count
+            start_count += 1
+
+        tasks = [Actor(duration=float("inf"), resources={"CPU": 1}, start_callback=start_callback) for _ in range(10000)]
+        simulator.submit(tasks)
+
+        time = 0
+        while start_count < len(tasks):
+            time = simulator.step()
+
+        assert time < 200
```

Comment:
Is it deterministic? Wondering if we can hardcode expected virtual time for all these tests.

Reply:
yup! it's deterministic, i figured it's better to leave some leeway though for the off-by-ones. For example, at t=0 the autoscaler runs before we schedule tasks on the cluster, so we have to be careful about reasoning about how long the simulation will take +-5s (the autoscaler's update frequency)

Reply:
Maybe we can assert within a range then, to ensure behaviour is reasonably pinned.

Reply:
sounds good, i will do that after the current launch rate limiting bug is fixed, because the times should drastically increase after that.

Reply:
Following up on @wuisawesome last comment here and @ericl comment (3) below. When we add this assert after the launch rate limiting bug is fixed, can we also assert the number of nodes to make sure we do not have any other bugs too?

Reply:
Perhaps I missed it, but where is the new fix?